rewrite.ml

```diff
@@ -0,0 +1,228 @@
+    Pulled_to (edit, pc_map pc_branch - 1)
+  end
+
+type progress = Made_progress | Stuck
+
+let move_up_instr is_eliminating is_anihilating is_blocking instrs pc =
+  let push_instr = push_instr is_eliminating is_anihilating is_blocking in
+  let rec work_push instrs progress to_push to_pull =
+    match to_push with
+    | pc :: to_push ->
+      begin match push_instr instrs pc with
+        | Blocked ->
+          work_push instrs progress to_push to_pull
+        | Stop (instrs, pc_map) ->
+          let (!!) = List.map pc_map in
+          work_push instrs progress !!to_push !!to_pull
+        | Need_pull pc' ->
+          work_push instrs progress to_push (pc' :: to_pull)
+        | Work ((instrs, pc_map), pcs) ->
+          let (!!) = List.map pc_map in
+          work_push instrs Made_progress (List.rev_append pcs !!to_push) !!to_pull
+      end
+    | [] -> work_pull instrs progress to_pull []
+  and work_pull instrs progress to_pull blocked_branches =
+    match to_pull with
+    | pc :: to_pull ->
+      begin match pull_instr instrs pc with
+        | Pulled_to ((instrs, pc_map), pc') ->
+          let (!!) = List.map pc_map in
+          work_push instrs Made_progress [pc'] (List.rev_append !!blocked_branches !!to_pull)
+        | Blocked ->
+          work_pull instrs progress to_pull (pc :: blocked_branches)
+      end
+    | [] ->
+      match progress with
+      | Stuck ->
+        None
+      | Made_progress ->
+        Some instrs
+  in
+  work_push instrs Stuck pc []
+
+let move_up is_target is_eliminating is_anihilating is_blocking instrs =
+    let move_up_instr = move_up_instr is_eliminating is_anihilating is_blocking in
+
+    (* Loop around and move all drops until none moves anymore *)
+    let rec work status instrs pc =
+      let at_end = Array.length instrs = pc in
+      if at_end then
+        match status with
+        | Stuck -> None
+        | Made_progress ->
+          let instrs = match work Stuck instrs 0 with
+            | Some instrs -> instrs
+            | None -> instrs
+          in
+          Some instrs
+      else if is_target instrs.(pc) then
+        begin match move_up_instr instrs [pc] with
+          | None -> work status instrs (pc+1)
+          | Some instrs -> work Made_progress instrs 0
+        end
+      else
+        work status instrs (pc+1)
+    in
+    work Stuck instrs 0
+
+
+module Drop = struct
+  let is_blocking var instr =
+    VarSet.mem var (required_vars instr)
+
+  let is_eliminating var instr =
+    match[@warning "-4"] instr with
+    | Assign (x, _) | Clear x -> x = var
+    | _ -> false
+
+  let is_anihilating var instr =
+    match[@warning "-4"] instr with
+    | Decl_const (x, _) | Decl_mut (x, _) -> x = var
+    | _ -> false
+
+  let move_up_var instrs var =
+    let is_target instr =
+      match[@warning "-4"] instr with
+      | Drop x when x = var -> true
+      | _ -> false
+    in
+    let move = move_up
```

Comment:
It's a bit confusing to have two different functions called `move_up`. Especially when it's **not** the function defined just a few lines later.